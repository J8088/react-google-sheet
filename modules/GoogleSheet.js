import React from 'react'
import PropTypes from 'prop-types'
import equalByKeys from '@lourd/equal-by-keys'
import { GoogleApiConsumer } from '@lourd/react-google-api'

class GSheetData extends React.Component {
  static propTypes = {
    id: PropTypes.string.isRequired,
    range: PropTypes.string.isRequired,
    api: PropTypes.object.isRequired,
  }

  state = {
    error: null,
    data: null,
    loading: false,
  }

  constructor(props) {
    super(props)
    this.fetch = this.fetch.bind(this)
  }

  componentDidMount() {
    this.fetch()
  }

  componentDidUpdate(prevProps) {
    if (!equalByKeys(this.props, prevProps, 'id', 'range')) {
      this.fetch()
    }
  }

  async fetch() {
    this.setState({ loading: true })
    try {
      const params = {
        spreadsheetId: this.props.id,
        range: this.props.range,
      }
      const response = await this.props.api.client.sheets.spreadsheets.values.get(
        params,
      )
      // Unable to cancel requests, so we wait until it's done and check it's still the desired one
      if (
        this.props.id === params.spreadsheetId &&
        this.props.range === params.range
      ) {
        this.setState({
          loading: false,
          error: null,
          data: response.result.values,
        })
      }
    } catch (response) {
      // If the api is still null, this will be a TypeError, not a response object
      const error = response.result ? response.result.error : response
      this.setState({ loading: false, error })
    }
  }

  render() {
    return this.props.children({
      error: this.state.error,
      data: this.state.data,
      loading: this.state.loading,
      refetch: this.fetch,
    })
  }
}

const GoogleSheet = props => (
  <GoogleApiConsumer>
    {api => <GSheetData api={api} {...props} />}
  </GoogleApiConsumer>
)

export default GoogleSheet
export { GSheetData }
