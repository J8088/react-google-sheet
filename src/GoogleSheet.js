import React from 'react';
import PropTypes from 'prop-types';
import { GoogleApiConsumer } from './GoogleApi';
import equalByKeys from './equalByKeys';

class GSheetData extends React.Component {
  static propTypes = {
    id: PropTypes.string.isRequired,
    range: PropTypes.string.isRequired,
    api: PropTypes.object.isRequired
  };

  state = {
    error: null,
    data: null,
    loading: false
  };

  componentDidMount() {
    this.fetch();
  }

  componentDidUpdate(prevProps) {
    if (!equalByKeys(this.props, prevProps, 'id', 'range')) {
      this.fetch();
    }
  }

  fetch = async () => {
    this.setState({ loading: true });
    try {
      const params = {
        spreadsheetId: this.props.id,
        range: this.props.range
      };
      const response = await this.props.api.client.sheets.spreadsheets.values.get(
        params
      );
      // Unable to cancel requests, so we wait until it's done and check it's still the desired one
      if (
        this.props.id == params.spreadsheetId &&
        this.props.range == params.range
      ) {
        this.setState({
          loading: false,
          error: null,
          data: response.result.values
        });
      }
    } catch (response) {
      this.setState({ loading: false, error: response.result.error });
    }
  };

  render() {
    return this.props.children({
      error: this.state.error,
      data: this.state.data,
      loading: this.state.loading,
      refetch: this.fetch
    });
  }
}

const GoogleSheet = props => (
  <GoogleApiConsumer>
    {api => <GSheetData api={api} {...props} />}
  </GoogleApiConsumer>
);

export default GoogleSheet;
