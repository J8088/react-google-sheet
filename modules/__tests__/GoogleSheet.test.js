import React from 'react'
import { mount } from 'enzyme'

import Deferred from '@lourd/deferred'
import { GoogleSheetApi, GoogleSheet } from '../'
import { GSheetData } from '../GoogleSheet'

it("renders the underlying GSheetData component with the api from context and passes through props, freaking out if there isn't an api provided in context", () => {
  jest.spyOn(console, 'error').mockImplementation(() => {})
  const fn = jest.fn(() => null)
  const props = {
    id: 'foo',
    range: 'things',
    foo: 'bar',
  }
  const result = mount(<GoogleSheet {...props}>{fn}</GoogleSheet>)
  expect(result).toMatchSnapshot('mounted component')
  expect(fn).toMatchSnapshot('child function calls')
  expect(console.error).toMatchSnapshot('logged errors')
  console.error.mockRestore()
})

describe('GSheetData component', () => {
  it('calls the api methods to load the data and renders the children function with the data loading state and a refetch function', async () => {
    const api = {
      client: {
        sheets: {
          spreadsheets: {
            values: {
              get: jest.fn(),
            },
          },
        },
      },
    }
    const deferred = new Deferred()
    api.client.sheets.spreadsheets.values.get.mockReturnValueOnce(
      deferred.promise,
    )
    const props = {
      id: 'foo',
      range: 'bar',
      api,
    }
    const children = jest.fn(() => null)
    const wrapper = mount(<GSheetData {...props}>{children}</GSheetData>)
    expect(wrapper.state()).toMatchSnapshot('loading state')
    expect(api.client.sheets.spreadsheets.values.get).toMatchSnapshot(
      'spreadsheet data loading method call',
    )
    const response = {
      result: {
        values: ['🤖', '👽', '💩'],
      },
    }
    deferred.resolve(response)
    // wait here for component updates to occur
    await deferred.promise
    expect(wrapper.state()).toMatchSnapshot('resolved state')
    expect(children).toMatchSnapshot('children function calls')
  })

  it('refetches if the id or range props changes', () => {
    const response = { result: { values: ['🤖', '👽', '💩'] } }
    const api = {
      client: {
        sheets: {
          spreadsheets: {
            values: { get: jest.fn(() => Promise.resolve(response)) },
          },
        },
      },
    }
    const props = { id: 'foo', range: 'bar', api }
    const wrapper = mount(<GSheetData {...props}>{() => null}</GSheetData>)
    wrapper.setProps({ id: 'different' })
    wrapper.setProps({ range: 'also different ' })
    expect(api.client.sheets.spreadsheets.values.get).toMatchSnapshot(
      'multiple data loading calls',
    )
  })

  it("doesn't update state if the id or range changed while the data was being fetched", async () => {
    const api = {
      client: {
        sheets: {
          spreadsheets: { values: { get: jest.fn() } },
        },
      },
    }
    const request1 = new Deferred()
    const request2 = new Deferred()
    api.client.sheets.spreadsheets.values.get
      .mockReturnValueOnce(request1.promise)
      .mockReturnValueOnce(request2.promise)
    const children = jest.fn(() => null)
    const response1 = { result: { values: ['🤖', '👽', '💩'] } }
    const response2 = { result: { values: ['💂', '👩‍🎤', '👢'] } }
    const props = { id: 'foo', range: 'bar', api }
    const wrapper = mount(<GSheetData {...props}>{children}</GSheetData>)
    wrapper.setProps({ id: 'different' })
    request1.resolve(response1)
    await request1.promise
    expect(wrapper.state()).toMatchSnapshot('final state')
  })

  it('sets the error in state if the data request has an error', async () => {
    const deferred = new Deferred()
    const api = {
      client: {
        sheets: {
          spreadsheets: { values: { get: jest.fn(() => deferred.promise) } },
        },
      },
    }
    const props = { id: 'foo', range: 'bar', api }
    const children = jest.fn(() => null)
    const wrapper = mount(<GSheetData {...props}>{children}</GSheetData>)
    const error = new Error('👻👹')
    const response = { result: { error } }
    deferred.reject(response)
    // have to wait on the promise here so rejecting it propagates through the react component update
    await expect(deferred.promise).rejects
    expect(wrapper.state()).toMatchSnapshot('resolved error state')
    expect(children).toMatchSnapshot('children function calls with error state')
  })
})
