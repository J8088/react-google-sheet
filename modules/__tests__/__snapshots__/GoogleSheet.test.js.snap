// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GSheetData component calls the api methods to load the data and renders the children function with the data loading state and a refetch function: children function calls 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": null,
        "error": null,
        "loading": false,
        "refetch": [Function],
      },
    ],
    Array [
      Object {
        "data": null,
        "error": null,
        "loading": true,
        "refetch": [Function],
      },
    ],
    Array [
      Object {
        "data": Array [
          "ðŸ¤–",
          "ðŸ‘½",
          "ðŸ’©",
        ],
        "error": null,
        "loading": false,
        "refetch": [Function],
      },
    ],
  ],
}
`;

exports[`GSheetData component calls the api methods to load the data and renders the children function with the data loading state and a refetch function: loading state 1`] = `
Object {
  "data": null,
  "error": null,
  "loading": true,
}
`;

exports[`GSheetData component calls the api methods to load the data and renders the children function with the data loading state and a refetch function: resolved state 1`] = `
Object {
  "data": Array [
    "ðŸ¤–",
    "ðŸ‘½",
    "ðŸ’©",
  ],
  "error": null,
  "loading": false,
}
`;

exports[`GSheetData component calls the api methods to load the data and renders the children function with the data loading state and a refetch function: spreadsheet data loading method call 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "range": "bar",
        "spreadsheetId": "foo",
      },
    ],
  ],
}
`;

exports[`GSheetData component doesn't update state if the id or range changed while the data was being fetched: final state 1`] = `
Object {
  "data": null,
  "error": null,
  "loading": true,
}
`;

exports[`GSheetData component refetches if the id or range props changes: multiple data loading calls 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "range": "bar",
        "spreadsheetId": "foo",
      },
    ],
    Array [
      Object {
        "range": "bar",
        "spreadsheetId": "different",
      },
    ],
    Array [
      Object {
        "range": "also different ",
        "spreadsheetId": "different",
      },
    ],
  ],
}
`;

exports[`GSheetData component sets the error in state if the data request has an error: children function calls with error state 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": null,
        "error": null,
        "loading": false,
        "refetch": [Function],
      },
    ],
    Array [
      Object {
        "data": null,
        "error": null,
        "loading": true,
        "refetch": [Function],
      },
    ],
    Array [
      Object {
        "data": null,
        "error": [Error: ðŸ‘»ðŸ‘¹],
        "loading": false,
        "refetch": [Function],
      },
    ],
  ],
}
`;

exports[`GSheetData component sets the error in state if the data request has an error: resolved error state 1`] = `
Object {
  "data": null,
  "error": [Error: ðŸ‘»ðŸ‘¹],
  "loading": false,
}
`;

exports[`renders the underlying GSheetData component with the api from context and passes through props, freaking out if there isn't an api provided in context: child function calls 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": null,
        "error": null,
        "loading": false,
        "refetch": [Function],
      },
    ],
    Array [
      Object {
        "data": null,
        "error": [TypeError: Cannot read property 'client' of null],
        "loading": false,
        "refetch": [Function],
      },
    ],
  ],
}
`;

exports[`renders the underlying GSheetData component with the api from context and passes through props, freaking out if there isn't an api provided in context: logged errors 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Warning: Failed prop type: The prop \`api\` is marked as required in \`GSheetData\`, but its value is \`null\`.
    in GSheetData (created by GoogleApiConsumer)
    in GoogleApiConsumer (created by GoogleSheet)
    in GoogleSheet (created by WrapperComponent)
    in WrapperComponent",
    ],
    Array [
      "Warning: <Consumer> was rendered outside the context of its <Provider>",
    ],
  ],
}
`;

exports[`renders the underlying GSheetData component with the api from context and passes through props, freaking out if there isn't an api provided in context: mounted component 1`] = `
<GoogleSheet
  foo="bar"
  id="foo"
  range="things"
>
  <GoogleApiConsumer
    quiet={false}
  >
    <GSheetData
      api={null}
      foo="bar"
      id="foo"
      range="things"
    />
  </GoogleApiConsumer>
</GoogleSheet>
`;
